src/python/pants/core/goals/build.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Optional
+
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.fs import Digest, MergeDigests, Snapshot, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import FieldSet, TargetRootsToFieldSets, TargetRootsToFieldSetsRequest
+from pants.engine.unions import union
+
+logger = logging.getLogger(__name__)
+
+
+@union
+class BuildFieldSet(FieldSet, metaclass=ABCMeta):
+    """The fields necessary to build an asset from a target."""
```

Comment:
It feels not quite right to me to document this as "the fields necessary to build", when there are no fields declared and it's abstract. Perhaps
```suggestion
    """The `@union` necessary to hook up a `./pants build` implementation from a target"""
```

Reply:
Check out the docs on how we present FieldSets to plugin authors: https://www.pantsbuild.org/docs/rules-api-and-target-api#fieldsets

While the superclass is an interface, this description is accurate for subclasses. In our plugin docs and the example plugin docs, we only show how you subclass a field set. We don't talk about creating the interface itself, as that's only relevant when you're creating a generic goal like `test`, `lint`, etc; at that advanced of a use case, we'd either expect users to read our code for inspiration, or to reach out to us for help.

Reply:
> Separately, we don't currently enforce e.g. that if a @union class has metaclass=ABCMeta, that all of its implementors must subclass it. Would that be a useful check to add? Right now, I believe it's possible to add UnionRule(BuildFieldSet, NoneType) without causing any "compile-time" error, for example. I believe the intent of this pattern has now become pretty canonical.

Yes, I think this would be useful. There are now several times where we use this pattern of the superclass being an ABC, and expecting subclasses to implement it.

If possible, it'd be great for an implementation to leverage the stdlib and ABCs. I believe there are mechanisms already in Python to check if a subclass inherits an ABC (it might be as simple as `isinstance`).

Possible implementation:

```python
base_is_an_interface = abc.is_abstract(union_base)  # I didn't check if this existed
invalid_members = [member for member in union_members if not issubclass(member, base)]
if invalid_members:
  raise Exception()
```

I do not think we need to support having this check be used if you are not using an ABC for your union base. It does not need to be more general than the above, imo.

Also, it's important that we continue to support a union base not being an ABC and the members not being subclasses. Namely, how we implement `PluginField`, where there is no inheritance relationship.